KEP-3726: standard-application-protocols

LiorLieberman opened the pull request:
<!-- 
	Please use the following format when naming your PR
	< Issue Number >:< Issue Description >
	e.g. KEP-000: adding beta graduation criteria
	
	Avoid using phrases like `fixes #NNNN` in the description
	unless the pull request is to change the KEP status to 
	implemented or KEP has been deprecated.
-->

<!-- short description of work done in PR e.g. updating milestone, adding new KEP, adding test requirements… -->  
- One-line PR description: adding new KEP for declaring standard application protocols
<!-- link to the k/enhancements issue -->
- Issue link: https://github.com/kubernetes/enhancements/issues/3726

<!-- other comments or additional information -->
- Other comments:
/sig network
/cc @aojea @howardjohn @robscott @bowei @danwinship

LiorLieberman commented on keps/sig-network/3726-standard-application-protocols/README.md:
```diff
@@ -0,0 +1,236 @@
+Those common protocols will be well defined strings prefixed with ‘k8s.io’. 
+`k8s.io/h2c` as an example.
+
+### list of protocols
+- 'k8s.io/h2c'
+- 'k8s.io/h2'
+- 'k8s.io/grpc'
+- 'k8s.io/grpc-web'
+- 'k8s.io/tls'
+- 'k8s.io/tcp'
```
Do we think all of them are needed?

Alternatively, are there protocols I missed?

robscott replied:
I think I'd drop `grpc-web` (not prevalent enough), and `tcp` (duplicative of port protocol). I'm also not sure how `tls` would differ in practice from `https` that's already defined as an IANA standard name, so I think it could make sense to leave that off as well.

howardjohn replied:
`tcp` is not duplicative.


HTTP:
```
appProtocol: k8s.io/HTTP
protocol: TCP
```

"Who knows, do whatever you want":
```
protocol: TCP
```


"This is definitely TCP"
```
appProtocol: k8s.io/TCP #might be IANA though, not sure prefix is needed
protocol: TCP
```

I know at least Istio and Linkerd treat these 3 cases differently

LiorLieberman replied:
You mean that here Istio tries to sniff and guess the app protocol 
```
protocol: TCP
```

and here it does not? 
```
appProtocol: k8s.io/TCP # it is unprefixed "tcp" in the current istio support
protocol: TCP
```

howardjohn replied:
Yes. if the duplicative "TCP" is confusing, you can also think of it as 

```
appProtocol: k8s.io/opaque-bytestream # it is unprefixed "tcp" in the current istio support
protocol: TCP
```

howardjohn commented on keps/sig-network/3726-standard-application-protocols/README.md:
```diff
@@ -0,0 +1,236 @@
+Those common protocols will be well defined strings prefixed with ‘k8s.io’. 
+`k8s.io/h2c` as an example.
+
+### list of protocols
```
Especially h2 vs h2c, and why we even need h2 when TLS already provides a mechanism for negotiating protocol

robscott replied:
In this case the relevant RFC includes clear definitions of both h2 and h2c (https://www.rfc-editor.org/rfc/rfc7540#section-3.1), I think it makes sense to copy those concepts and link out to the RFC. Although there is a mechanism for negotiating protocol, at least some implementations also provide a way to explicitly specify h2 or h2c.

howardjohn replied:
I guess a broader topic is - what happens when there is limited support.

For example, Istio cannot implement something that says "This is HTTPS, but it must be HTTP/2". But we could do "HTTPS", which is probably good enough for 99% of the users declaring "H2"... but then for that 1% of users, we didn't meet their demand (used HTTP/1 when they wanted h2). Alternatively, we could decide "This is H2, we can't always give that behavior, so we ignore it".

Hope that makes sense...

robscott replied:
In my opinion this is a good reason to make this a list of protocols supported by the underlying application. That feels like a follow up though. The scope of this KEP is really just trying to fill gaps in IANA Standard Service names which feels like it should be relatively non-controversial, even though we admit that follow up efforts will be needed to solve the larger problem.

thockin replied:
I didn't follow the example or the argument for a list, can you do it again but slower?

LiorLieberman replied:
I think I caused a confusion. 

This line is **"List of supported/standard protocols"** and it is **not** about changing the field to a list (which is suggested later in the followup work section)

I will rename this line to supported protocols.

To answer your question @thockin ; The idea is to declare standard protocol names for common protocols that are not IANA service names. That way in the current version of the field, implementations (istio/GKE etc etc) that implement non IANA service names like `http2` will be able to coexist. (without these standards they would need `gke.io/http2` and `istio.io/http2` and since `AppProtocol` is a string field now it can have only one of the options)

Probably for that reason most of the implementations support non prefixed app protocol names that are not IANA (and that should not be the case) - [istio as an example](https://github.com/istio/istio/blob/master/pkg/config/protocol/instance.go#L65-L94)

There is a suggested followup work to turn this field to a list instead of a string and truly support implementations to interoperate. (lets say two different impl specific prefixed protocols OR one impl specific prefixed and one not prefixed etc)

LiorLieberman commented:
I am going to add tomorrow to the KEP the relevant change I am proposing to the godoc in the two places AppProtocol is in which is [ServicePort](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L4080-L4087) and [EndpointPort](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/discovery/types.go#L162-L169).

But ideally it will be the list of protocols in the KEP along with their meaning

robscott commented on keps/sig-network/3726-standard-application-protocols/README.md:
```diff
@@ -0,0 +1,236 @@
+- 'k8s.io/h2c'
+- 'k8s.io/h2'
+- 'k8s.io/grpc'
+- 'k8s.io/grpc-web'
```
I agree that grpc-web is probably not prevalent enough to be a part of this list, so probably worth pulling that out.

With that said, I think it's important to distinguish that these app protocols represent concepts that are recognized but not implemented by Kubernetes. AppProtocol exists as a hint to implementations adding functionality on top of the Kubernetes Service spec. I look at this KEP as defining terms that we expect will be commonly used and broadly understood but are not already defined as an IANA service name.

danwinship replied:
> Even gRPC - do we need to distinguish from h2?

The API docs say:

```
	// The application protocol for this port.
	// This field follows standard Kubernetes label syntax.
	// Un-prefixed names are reserved for IANA standard service names (as per
	// RFC-6335 and https://www.iana.org/assignments/service-names).
	// Non-standard protocols should use prefixed names such as
	// mycompany.com/my-custom-protocol.
```

The user docs say:

> The `appProtocol` field provides a way to specify an application protocol for each Service port. The value of this field is mirrored by the corresponding Endpoints and EndpointSlice objects.
> 
> This field follows standard Kubernetes label syntax. Values should either be [IANA standard service names](https://www.iana.org/assignments/service-names) or domain prefixed names such as `mycompany.com/my-custom-protocol`.

So the answer to your question is: **???** None of the docs explain what the field actually _for_, or who is expected to use it, for what, so how could we say whether it's important to distinguish grpc from generic h2?

aojea commented on keps/sig-network/3726-standard-application-protocols/README.md:
````diff
@@ -0,0 +1,236 @@
+
+```Values should either be IANA standard service names or domain prefixed names such as mycompany.com/my-custom-protocol.```
+
+This KEP proposes to declare standard Kubernetes protocol names for common protocols that are not IANA standard service names.
````
where is going to live this list?

aojea replied:
I think that the KEP should answer the where, who and how this list is going to be created, maintained, ... what are the requirements ... or we can be bikeshedding for years for each protocol

robscott replied:
I think the list is small enough that it could live in the spec for AppProtocol, but open to alternatives

aojea replied:
Do you keep the list versioned matching Kubernetes releases? i.e. the list in 1.26 only applies to 1.26 or the previous one too?

robscott replied:
Yeah I don't think there's any reason this needs to be versioned, but I also can't think of any other place this could fit. Agree that the KEP needs to define the requirements for protocols that get this status.

LiorLieberman replied:
If I understand correctly(please correct me if not), you are suggesting to have a new repo with standard names and/or structured metadata in it for various things in k8s.

While this can be a good idea, how will we use it for this case? 
We cannot validate the `AppProtocol` field name against it because we are not aware of whether the provided string for this field is an IANA service name, custom impl-specific prefixed name OR something unsupported.

maybe we can reference to the relevant parts of this registry from the field description or from the docs though.

aojea replied:
are we diligent enough to maintain another repo? 
how do we "enforce" devs to register everything in the new repo?
it looks to me that the idea is good but it got outdated soon, I can see how docs and release folks do a lot of work chasing people to update things.
I like the idea of being in-tree and use some of your "makefile" magic @thockin , that also will allow to be in ` importable Go form`

thockin replied:
I'll slap together a basic KEP

thockin replied:
https://github.com/kubernetes/enhancements/pull/3787

aojea commented:
/ok-to-test

LiorLieberman commented:
/test pull-enhancements-verify

LiorLieberman commented:
/test pull-enhancements-verify

LiorLieberman commented:
/test pull-enhancements-verify

aojea commented:
@sftim may have some good ideas about this too, I'd be interested in his thoughts

danwinship commented on hack/verify-toc.sh:
```diff
@@ -47,6 +47,7 @@ echo "Checking table of contents are up to date..."
     | xargs mdtoc --inplace --max-depth=5 --dryrun || (
-      echo "Table of content not up to date. If this failed silently and you are on mac, try 'brew install grep'"
+      echo "Table of content not up to date. Did you run 'make update-toc'?"
+      echo "If this failed silently and you are on mac, try 'brew install grep'"
```
split this out as a separate PR?

LiorLieberman replied:
Thanks, fixed it. it got in by accident - I already opened another PR for that.

danwinship commented on keps/sig-network/3726-standard-application-protocols/README.md:
```diff
@@ -0,0 +1,276 @@
+* Not be an [IANA standard service name](https://www.iana.org/assignments/service-names)
+* Run on top of L4 protocol supported by Kubernetes Service
+* Be supported in more than three implementations
+* Be well defined and broadly used
```
So this does nothing to stop us from running into the "gke `HTTP2` vs Istio `HTTP2`" problem for new protocols in the future; we have to get three possibly-incompatible implementations first before we're allowed to fix things...

Adding `k8s.io/h2` and `k8s.io/h2c` solves the _specific_ problem that led to this KEP, but we're still leaving the foot gun fully armed for future users.

LiorLieberman replied:
If implementations follow the docs they should not support unprefixed protocols that are not IANA service names. That means that should do "istio.io/<protocol>" or gke.io/<protocol> if the protocol is not IANA standard name. 

I agree that 3 may not be ideal, also the domain prefixed protocol make it impossible for implementations to coexist and support not IANA standard name. For that @robscott suggested future work to turn this field to a list.

When this happens, implementations could truly interoperate. However the field name is no longer appropriate (it is debatable if it was before) since it is not longer the "application protocol" it is more of a hint to the implementations on how to forward the traffic to the application.

robscott replied:
Yeah I think the ultimate solution is going to be a list that supports multiple values here since there's probably always going to be some implementation specific values.

thockin replied:
Would it be sufficient to say "at least two in-progress implementations of components which would use this protocol" ?

LiorLieberman replied:
changed it to two or more. 
Does "in progress" is enough for us? I think we'd want to add another standard protocol only when we know there is a real need for it and it is fully supported

LiorLieberman replied:
Yeah, though until it is a list I see Tim's and Dan's point where users will use istio.io/custom or something else and then they would need to migrate to k8s.io
And implementation will need to support both for backward compatibility.

aojea replied:
if we define the field as an url we can also apply url rules to implement compatibility.  
We can use subdomains to define implementations, and paths to define protocols , it can also allow to do composition for protocols that tunnels  `<implementation>.k8s.io/<protocol1>/<protocol>`

Per example 

```
istio.k8s.io/http/v2
```

will mean the http v2 implementation from istio, when the standard one will be `k8s.io/http/v2`

Something more complex

```
gateway.k8s.io/websocket/sip
```

that means the implementation of SIP over Websocket (I just learned about this recently, no idea it was a thing :)

thockin replied:
Would Istio and GKE even qualify for the "two implementations" rule?  They used the same name but for different meanings.  If they had both followed the rules and used their own prefixes, we would only have 1 implementation of h2c and one of h2, right?

Perhaps add this as an UNRESOLVED and let's get this merged so we can debate the fine points.

LiorLieberman replied:
Agreed, I have added this as UNRESOLVED.

LiorLieberman commented:
@robscott @howardjohn @aojea @danwinship PTAL

robscott commented on keps/sig-network/3726-standard-application-protocols/README.md:
```diff
@@ -0,0 +1,284 @@
+
+## Summary
+
+There are cases where implementations implement different things for the same application protocol names. That already causes(and can cause more in the future) issues for certain implementations to interoperate. (see example with GKE and Istio under [Motivation](#motivation))
```
Tiny Nits:
```suggestion
There are cases where implementations implement different things for the same application protocol names. That has already caused issues for certain implementations to interoperate, with the possibility of more in the future. (See the example with GKE and Istio under [Motivation](#motivation))
```

robscott commented on keps/sig-network/3726-standard-application-protocols/README.md:
```diff
@@ -0,0 +1,284 @@
+
+## Motivation
+
+The lack of direct support for specifying application protocols for ports and the widespread use of implementation-specific annotations to mitigate it had led k8s to add the `AppProtocol` field to the port spec.
```
```suggestion
The lack of direct support for specifying application protocols for ports and the widespread use of implementation-specific annotations to mitigate it had led Kubernetes to add the `AppProtocol` field to the port spec.
```

robscott commented on keps/sig-network/3726-standard-application-protocols/README.md:
```diff
@@ -0,0 +1,284 @@
+
+The lack of direct support for specifying application protocols for ports and the widespread use of implementation-specific annotations to mitigate it had led k8s to add the `AppProtocol` field to the port spec.
+
+While this is a good solution - we never came with recommended standards other than [IANA standard service names](https://www.iana.org/assignments/service-names) or a domain prefixed protocol.
```
```suggestion
While this is a good solution - [IANA standard service names](https://www.iana.org/assignments/service-names) do not define some very common protocols.
```

robscott commented on keps/sig-network/3726-standard-application-protocols/README.md:
```diff
@@ -0,0 +1,284 @@
+
+## Design Details
+
+At first, the list is going to live in `ServicePort` and `EndpointPort` as part of the AppProtocol description.
+
+Based on the list size, we might revisit this decision in the future and suggest an alternative location.
```
I think we should remove any changes to a list from the scope of this KEP.

LiorLieberman replied:
List as list of supported protocols not as turning AppProtocol to a list

LiorLieberman replied:
as per Antonio's comment here https://github.com/kubernetes/enhancements/pull/3727#discussion_r1068776937

LiorLieberman replied:
changed the wording to avoid confusion

robscott commented on keps/sig-network/3726-standard-application-protocols/README.md:
```diff
@@ -0,0 +1,284 @@
+In order to be included in the list, a new protocol must:
+* Not be an [IANA standard service name](https://www.iana.org/assignments/service-names)
+* Run on top of L4 protocol supported by Kubernetes Service
+* Be supported in more than three implementations
```
Maybe we should loosen this requirement
```suggestion
* Be supported in two or more implementations
```

thockin replied:
implementations of ... ?

robscott commented on keps/sig-network/3726-standard-application-protocols/README.md:
```diff
@@ -0,0 +1,284 @@
+### Followup work
+To support implementations interoperability with different domain prefixed protocols (or a mix domain prefixed and non prefixed protocol) for the same port we need to turn `AppProtocol` to a list.
+
+It is likely to be an API change but design details TBD.
+### Documentation change
```
```suggestion
It is likely to be an API change but design details TBD.

### Documentation change
```

robscott commented on keps/sig-network/3726-standard-application-protocols/kep.yaml:
```diff
@@ -0,0 +1,35 @@
+creation-date: 2023-01-11
+reviewers:
+  - "@aojea"
+  - "@robscott"
+  - "@robscott"
```
😆 
```suggestion
  - "@robscott"
```

robscott reviewed (commented):
Thanks @LiorLieberman!

thockin commented on keps/sig-network/3726-standard-application-protocols/README.md:
```diff
@@ -0,0 +1,284 @@
+
+## Design Details
+
+At first, the list is going to live in `ServicePort` and `EndpointPort` as part of the AppProtocol description.
```
What is "the list" ?

thockin replied:
Oh, the list of protocols?  Let's jusmp straight to metadata.

LiorLieberman replied:
I have changed the wording a bit

LiorLieberman commented on keps/sig-network/3726-standard-application-protocols/README.md:
```diff
@@ -0,0 +1,284 @@
+2. Have no support for implementations interoperability with different domain prefixed protocols (or a mix domain prefixed and non prefixed protocol) for the same port.
+
+One good example for the first bullet is `HTTP2`. 
+* In GKE you can use `appProtocol: HTTP2` and it will describe HTTP2 over TLS (https://cloud.google.com/kubernetes-engine/docs/how-to/secure-gateway#load-balancer-tls).
```
Correct ideally that should not be used by anyone, but the fact is that different implementations are using it.
I have provided more context here - https://github.com/kubernetes/enhancements/pull/3727#discussion_r1087712029

LiorLieberman commented:
/test pull-enhancements-verify

thockin commented:
Is this ready for re-review?

On Fri, Jan 27, 2023 at 11:11 AM Rob Scott ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In keps/sig-network/3726-standard-application-protocols/README.md
> <https://github.com/kubernetes/enhancements/pull/3727#discussion_r1089327445>
> :
>
> > +  //
> +  // * Other protocols should use prefixed names such as
> +  // mycompany.com/my-custom-protocol.
> +  // +optional
> +  AppProtocol *string
> +```
> +
> +same wording for type `EndpointPort`
> +
> +### Adding new protocols
> +
> +In order to be included in the list, a new protocol must:
> +* Not be an [IANA standard service name](https://www.iana.org/assignments/service-names)
> +* Run on top of L4 protocol supported by Kubernetes Service
> +* Be supported in more than three implementations
> +* Be well defined and broadly used
>
> We do still leave room open for domain prefixed values here, so if we
> don't have 2 or more implementations, that's probably the way to go. I also
> think we need to expand this to a list in a future KEP so we can actually
> allow multiple impl-specific values here (ie istio.io/custom-protocol and
> linkerd.io/other-protocol).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/enhancements/pull/3727#discussion_r1089327445>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKWAVBMBIBAAIIZ4TCB3N3WUQMWPANCNFSM6AAAAAATYNACV4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

LiorLieberman commented:
Yes,
I have not looked at the KEP you have opened
https://github.com/kubernetes/enhancements/pull/3787 but presumably we will
see if we should move values from here to the new Names DB in the future.
Unless we want to hold on this KEP until the Names DB is set up.

The other open bullet is the "in-progress" vs "already implemented" for the
requirement to add new standard protocols in the future.
I do not have a strong opinion here - I tend to agree with the
"in-progress" approach because it might take some time until we change this
field to a list.

On Fri, 27 Jan 2023 at 20:26, Tim Hockin ***@***.***> wrote:

> Is this ready for re-review?
>
> On Fri, Jan 27, 2023 at 11:11 AM Rob Scott ***@***.***> wrote:
>
> > ***@***.**** commented on this pull request.
> > ------------------------------
> >
> > In keps/sig-network/3726-standard-application-protocols/README.md
> > <
> https://github.com/kubernetes/enhancements/pull/3727#discussion_r1089327445
> >
> > :
> >
> > > + //
> > + // * Other protocols should use prefixed names such as
> > + // mycompany.com/my-custom-protocol.
> > + // +optional
> > + AppProtocol *string
> > +```
> > +
> > +same wording for type `EndpointPort`
> > +
> > +### Adding new protocols
> > +
> > +In order to be included in the list, a new protocol must:
> > +* Not be an [IANA standard service name](
> https://www.iana.org/assignments/service-names)
> > +* Run on top of L4 protocol supported by Kubernetes Service
> > +* Be supported in more than three implementations
> > +* Be well defined and broadly used
> >
> > We do still leave room open for domain prefixed values here, so if we
> > don't have 2 or more implementations, that's probably the way to go. I
> also
> > think we need to expand this to a list in a future KEP so we can actually
> > allow multiple impl-specific values here (ie istio.io/custom-protocol
> and
> > linkerd.io/other-protocol).
> >
> > —
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/kubernetes/enhancements/pull/3727#discussion_r1089327445
> >,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/ABKWAVBMBIBAAIIZ4TCB3N3WUQMWPANCNFSM6AAAAAATYNACV4
> >
> > .
> > You are receiving this because you were mentioned.Message ID:
> > ***@***.***>
> >
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/enhancements/pull/3727#issuecomment-1407039867>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASLANQRAYS3VSSQVQO2E2YTWUQVPDANCNFSM6AAAAAATYNACV4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

thockin commented:
> Unless we want to hold on this KEP until the Names DB is set up.

No need

thockin reviewed (commented):
@johnbelamaric Does this need a PRR file if it has no PRR (no code) ?

LiorLieberman commented:
Thanks @thockin, 
I have addressed all your feedback - regarding the two last bullets in the checklist; I need to change the documentation - https://kubernetes.io/docs/concepts/services-networking/service/#application-protocol 

should we merge this first and then I change it or there it should work differently?

thockin reviewed (commented):
Provisional - let's make sure to close last issues!

/lgtm
/approve
